File: memento-pattern/text-editor/text-editor.ts
// Memento: Stores the state of the editor.
class Memento {
  private state: string;

  constructor(state: string) {
    this.state = state;
  }

  getState(): string {
    return this.state;
  }
}

// Originator: The object whose state needs to be saved and restored.
class TextEditor {
  private content: string = "";

  write(text: string): void {
    this.content = text;
  }

  save(): Memento {
    // Saves the current state in a memento
    return new Memento(this.content);
  }

  restore(memento: Memento): void {
    // Restores state from a memento
    this.content = memento.getState();
  }

  read(): string {
    return this.content;
  }
}

// Caretaker: manages mementos
const editor = new TextEditor();

console.log("Initial content: ", editor.read());

// Let's write some text
editor.write(
  "Node.js can now run Typescript files natively using an experimental flag."
);
const savedState = editor.save();
console.log("Content after writing: ", editor.read());

// Let's modify the content now
editor.write(
  "But Bun supports typescript without an experimental flag. and what about Deno ?"
);
console.log("Content after modifying: ", editor.read());

// Restoring prev content
editor.restore(savedState);
console.log("Content after restoring: ", editor.read());

File: memento-pattern/text-editor/text-editor.rs
// Memento: Stores the state of the text editor
struct Memento {
    state: String,
}

impl Memento {
    fn new(state: &str) -> Self {
        Self {
            state: state.to_string(),
        }
    }

    fn get_state(&self) -> &str {
        &self.state
    }
}

// Originator: Manages the state of the text editor
struct TextEditor {
    content: String,
}

impl TextEditor {
    fn new() -> Self {
        Self {
            content: String::new(),
        }
    }

    fn write(&mut self, text: &str) {
        self.content = text.to_string();
    }

    fn save(&self) -> Memento {
        // Saves the current state in a memento
        Memento::new(&self.content)
    }

    fn restore(&mut self, memento: &Memento) {
        self.content = memento.get_state().to_string();
    }

    fn read(&self) -> &str {
        &self.content
    }
}

// Caretaker: Manages the mementos
fn main() {
    let mut editor = TextEditor::new();
    println!("Initial content: {}", editor.read());

    // Write and save text
    editor.write("Rust is awesome");
    let saved_state = editor.save();
    println!("Content after writing: {}", editor.read());

    // Modify content
    editor.write("Rust fixes C++ for beginners");
    println!("Content after modifying: {}", editor.read());

    // Restore prev content
    editor.restore(&saved_state);
    println!("Content after restoring: {}", editor.read());
}
